docs/enhancements/0002-container-runtime.md

```diff
@@ -0,0 +1,122 @@
+
+The current implementation of KMM is checking whether images exists by generating an HTTP[s] request to the
+container registry. It requires KMM to manually access a bunch of configurations files on the nodes such as:
+* Container images mirroring configs
+* TLS configs
+* Installed CAs
+
+In addition to that, in order to reduce the number of registy accesses, we also need to maintain a cache as
+we do for the Hub&Spoke topology.
+
+[We recently started using the cluster's container-runtime for pulling kmod images from the worker pods](https://github.com/kubernetes-sigs/kernel-module-management/pull/886) which was a great step forward to maximum automation for pulling images
+in the cluster. Once we start checking the images existance using the container-runtime as well, we will be
+at ~100% automation for handlering images - everything the cluster does when pulling images, KMM will do as well.
+
+## Goals
+
+1. Reduce the number of manuall actions KMM has to take to pull images correctly.
+
+## Non-goals
+
+1. Add further control for building/signing kmod images.
+
+## Design
+
+### The new `ModuleImagesConfig` CRD
+
+The new CRD will hold information regarding the existance of images mentioned in `Module` objects.
+The cluster will contain a single `ModuleImagesConfig` per `Module` in the cluster.
+
+```yaml
+apiVersion: kmm.sigs.x-k8s.io/v1beta1
+kind: ModuleImagesConfig
+metadata:
+  ...
+spec:
+  items:
+    - image: registry.example.com/org/kmod1:kver
+    - image: registry.example.com/org/kmod2:kver
+      build:
+        ....
+      sign:
+        ...
+    - image: registry.example.com/org/kmod3:kver
+status
+  items:
+    - image: registry.example.com/org/kmod1:kver
+      status: Exists
+    - image: registry.example.com/org/kmod2:kver
+      status: DoesNotExist
+```
+
+### Module-NMC controller's changes
+
+* The only controller creating `ModuleImagesConfig` objects - one per `Module`
+* Checking `ModuleImagesConfig`'s status to determine whether an `NMC` object should be created
+
+```mermaid
+flowchart TD
+    MOD((Module)) -->|Reconcile| MNMCC(Module-NMC controller)
+    MNMCC --> |No MIC exists for the Module| MIC((MIC)) --> DONE
+    MNMCC --> |Image's NOT in MIC's spec| A1{{Add it to ModuleImagesConfig's spec}} --> DONE(((Done)))
+    MNMCC --> |Image's in MIC's spec| J1((.))
+        J1 --> |Image has no status or status='DoesNotExist'| DONE
+        J1 --> |Image's status='Exists'| NMC((NMC)) --> DONE
+```
+
+### ModuleImagesConfig controller (MIC) (previously called the BuildSign controller) changes
+
+* Continue to build/sign kmods as before
+* Creating pods to check if images exists and updates the `ModuleImagesConfig`'s status accordingly
+
+**The following chart only describe the new functionality added to the controller.**
+```mermaid
+flowchart TD
+    MIC((ModuleImagesConfig)) -->|Reconcile| MICC(MIC controller)
+        MICC --> |Image has not status| POD((Pod))
```

Comment:
In case image has no status - NMC creates the pod and exits, it should not wait for pod completion. Once the pod is failed/completed, the reconcile loop will be kicked in again

Reply:
I will change it in the chart.